### PR TITLE
[LLaDA 2.0-Uni] Add LLaDA 2.0-Uni multimodal discrete diffusion pipeline

### PR DESCRIPTION
**Pull request by ChinChyi**
## What does this PR do?

Adds support for [LLaDA 2.0-Uni](https://huggingface.co/inclusionAI/LLaDA2.0-Uni), a unified multimodal discrete diffusion language model that supports text understanding, image understanding, and image generation in a single framework.

**Paper**: [LLaDA2.0-Uni: Unifying Multimodal Understanding and Generation with Diffusion Large Language Model](https://arxiv.org/abs/2604.20796)

### New Components

- **`LLaDA2UniImageTransformer2DModel`** — Image diffusion transformer for decoding VQ tokens to images
- **`UniLLaDaPipeline`** — Unified pipeline supporting three modes:
  - Text-to-image generation
  - Image understanding (VQA, captioning)
  - Image editing
- **`LLaDA2UniFlowMatchEulerScheduler`** — Flow matching scheduler with Euler ODE integration
- **Image tokenizer utilities** — SigVQ-based image encoding/decoding

### Key Features

- Multimodal capabilities: Single model handles both vision and language tasks
- Discrete diffusion: Block-wise iterative refinement for token generation
- FP8 quantization support: Efficient inference with quantized weights
- Flexible decoding: Supports both quality mode (50 steps) and turbo mode (8 steps)

### Usage Example

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
from diffusers import UniLLaDaPipeline, BlockRefinementScheduler
from diffusers.pipelines.unillada.image_tokenizer import ImageTokenizer

model_id = "inclusionAI/LLaDA2.0-Uni"
model = AutoModelForCausalLM.from_pretrained(
    model_id, trust_remote_code=True, torch_dtype=torch.bfloat16, device_map="auto"
)
tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True)
scheduler = BlockRefinementScheduler()
image_tokenizer = ImageTokenizer(model_path=model_id)

pipe = UniLLaDaPipeline(
    transformer=model,
    tokenizer=tokenizer,
    scheduler=scheduler,
    image_tokenizer=image_tokenizer,
)

# Text-to-Image
result = pipe(prompt="A cat sitting on a windowsill at sunset")
result.images[0].save("output.png")

# Image Understanding
from PIL import Image
img = Image.open("photo.jpg")
result = pipe(image=img, question="Describe this image in detail.")
print(result.text)

# Image Editing
result = pipe(image=img, instruction="Change the background to a beach.")
result.images[0].save("edited.png")
```

### Testing

- Added unit tests in `tests/pipelines/unillada/test_unillada.py`
- Tests cover all three modes (generation, understanding, editing)
- Mock components for CI compatibility

### Model Weights

Official weights available at: https://huggingface.co/inclusionAI/LLaDA2.0-Uni

## Before submitting

- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md)?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

@yiyixuxu @a-r-r-o-w @DN6

### CODE REVIEW

File: src/diffusers/models/transformers/transformer_llada2uni_image.py [hunk 1]
```diff
@@ -0,0 +1,1082 @@
...
+                self.freqs_cis = [freqs_cis.to(device) for freqs_cis in self.freqs_cis]
+
+        result = []
+        for i in range(len(self.axes_dims)):
+            index = ids[:, i]
+            result.append(self.freqs_cis[i][index])
+        return torch.cat(result, dim=-1)
+
+
+class LLaDA2UniImageTransformer2DModel(ModelMixin, ConfigMixin, PeftAdapterMixin, FromOriginalModelMixin):
```
Comment (dg845) [inline src/diffusers/models/transformers/transformer_llada2uni_image.py:400]: Is `LLaDA2UniImageTransformer2DModel` intended to be used as part of the UniLLaDA pipeline? I see that the `transformer` loaded by the pipeline is a remotely implemented `transformers` model (`LLaDA2MoeModelLM` in [`modeling_llada2uni_moe.py`](https://huggingface.co/inclusionAI/LLaDA2.0-Uni/blob/main/modeling_llada2uni_moe.py)), and this transformer doesn't appear to be used anywhere.

File: src/diffusers/pipelines/unillada/pipeline_unillada.py [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+        ```python
+        >>> import torch
+        >>> from transformers import AutoModelForCausalLM, AutoTokenizer
+        >>> from diffusers import UniLLaDaPipeline, BlockRefinementScheduler
+        >>> from diffusers.pipelines.unillada.image_tokenizer import ImageTokenizer
+
+        >>> model_id = "inclusionAI/LLaDA2.0-Uni"
+        >>> model = AutoModelForCausalLM.from_pretrained(
+        ...     model_id, trust_remote_code=True, torch_dtype=torch.bfloat16, device_map="auto"
+        ... )
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/pipeline_unillada.py:43]: I would prefer a `diffusers`-native (or `transformers`-native) implemention of the DiT model so that we don't need `trust_remote_code=True`.

File: src/diffusers/pipelines/unillada/pipeline_unillada.py [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+        block_length: int,
+        decoder_steps: int,
+        decode_mode: str,
+        decode_fn: Callable | None,
+        resolution_multiplier: int,
+        output_type: str,
+        return_dict: bool,
+    ) -> UniLLaDaPipelineOutput | tuple:
+        """Text-to-image generation."""
+        result = self.transformer.generate_image(
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/pipeline_unillada.py:408]: I think the denoising loop should be implemented in `UniLLaDaPipeline.__call__` using a scheduler (such as `BlockRefinementScheduler`), which is the standard `diffusers` design, rather than in transformer methods like `generate_image`.

File: src/diffusers/pipelines/unillada/image_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+    vqmodel.load_state_dict(vq_sd, strict=False)
+    del visual_sd, vq_sd
+
+
+# ============================================================
+# Main tokenizer class
+# ============================================================
+
+
+class ImageTokenizer:
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/image_tokenizer.py:513]:
```suggestion
class ImageTokenizer(ModelMixin, ConfigMixin):
```
I think `ImageTokenizer` should inherit from `ModelMixin` and `ConfigMixin` (which is standard for `diffusers` models) so that saving and loading can be handled in the normal `diffusers` way, rather needing to implement it separately in `__init__` below.

File: src/diffusers/pipelines/unillada/image_tokenizer.py [hunk 5]
```diff
@@ -0,0 +1,617 @@
...
+
+# ============================================================
+# Image preprocessing
+# ============================================================
+
+OPENAI_CLIP_MEAN = [0.48145466, 0.4578275, 0.40821073]
+OPENAI_CLIP_STD = [0.26862954, 0.26130258, 0.27577711]
+
+
+class ImagePreprocessor:
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/image_tokenizer.py:92]: I think we should refactor the image preprocessing logic in `ImagePreprocessor` into a dedicated `VaeImageProcessor` subclass that lives in its own file (e.g. `image_processor.py`). See for example `JoyImageEditImageProcessor` as a reference:

https://github.com/huggingface/diffusers/blob/68a4847768c9a4e5e39307635ff2762ef2ef5d13/src/diffusers/pipelines/joyimage/image_processor.py#L66

File: src/diffusers/pipelines/unillada/image_tokenizer.py [hunk 6]
```diff
@@ -0,0 +1,617 @@
...
+        # Try to use the HF attention dispatch (flash_attention_2 / sdpa / eager)
+        try:
+            from transformers.modeling_utils import ALL_ATTENTION_FUNCTIONS
+
+            attn_impl = getattr(self.config, "_attn_implementation", "eager")
+            if attn_impl != "eager" and attn_impl in ALL_ATTENTION_FUNCTIONS:
+                attention_interface = ALL_ATTENTION_FUNCTIONS[attn_impl]
+                if "flash" in attn_impl:
+                    max_seqlen = (cu_seqlens[1:] - cu_seqlens[:-1]).max()
+                    attn_output, _ = attention_interface(
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/image_tokenizer.py:240]: We should consider using `dispatch_attention_fn` instead as it handles the attention backends used here, such as Flash Attention (including `flash_varlen`) and `torch` native SDPA. For reference, see the [attention backend docs](https://huggingface.co/docs/diffusers/v0.38.0/en/optimization/attention_backends).

File: src/diffusers/pipelines/unillada/sigvq.py [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+            _LinearWrapper(dim, hidden_dim),
+            nn.SiLU(),
+            nn.Linear(hidden_dim, dim),
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.net(x)
+
+
+class SigVQ(nn.Module):
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/sigvq.py:33]: Is the `SigVQ` model intended to be used as part of the UniLLaDA pipeline? I don't see it being used anywhere.

File: src/diffusers/pipelines/unillada/utils.py [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+"""Image preprocessing utilities for UniLLaDA pipeline."""
+
+from __future__ import annotations
+
+import math
+
+import PIL.Image
+
+
+def generate_crop_size_list(
```
Comment (dg845) [inline src/diffusers/pipelines/unillada/utils.py:24]: Similar to https://github.com/huggingface/diffusers/pull/13686#discussion_r3247497883, I think we should refactor the image preprocessing logic here into a dedicated `VaeImageProcessor` subclass (possibly combined with the one from `image_tokenizer.py`).

**Review by dg845 — COMMENTED** [review]
Thanks for the PR! I left an initial design review :).

File: src/diffusers/pipelines/unillada/pipeline_unillada.py (hunk 2 above)
Comment (yiyixuxu) [reply src/diffusers/pipelines/unillada/pipeline_unillada.py:43]: it's really ok to have trust_remote_code here, ideally transformer native but it is up to them 
we are not going to host transformer models in diffusers

**Comment by dg845** [comment]
Hi @ChinChyi, a gentle ping on this. Happy to help out with anything :).
